0000-friends_in_high_kindednesses.md

```diff
@@ -0,0 +1,442 @@
+fn foo<T>(iter: T) where T: StreamingIterator, for<'a> T::Item<'a>: Display { ... }
+```
+
+This RFC does not propose allowing any sort of bound by the type constructor
+itself, whether an equality bound or a trait bound (trait bounds of course are
+also impossible). 
+
+#### `let` introduction of parameters in bound
+
+The `for` syntax for HRTB is widely considered inaccessible and difficult to learn.
+The problem here is not that the underlying concept of a higher rank parameter is
+particularly challenging, but that the syntax introduces too much jargony syntax
+which distracts from the underlying idea.
+
+This RFC proposes adding a new syntax for higher rank parameters which appear in the
+type position of the bound. A user can simply introduce a new parameter with the `let`
+keyword:
+
+```rust
+where T: Iterable, T::Item<let 'a>: &'a str
+//equivalent to
+where T: Iterable, for<'a> T::Item<'a>: &'a str
+```
+
+The variable introduced by the `let` is scoped only to this bound. Shadowing existing
+type variables in scope is not permitted by a `let`, just as it is not permitted with
+the existing `for` syntax. The `let` keyword is necessary to make it unambiguous that
+the user intends to introduce a new variable here.
+
+The `let` syntax is valid for any type constructor being bound, including those which
+are not associated items. As an arbitrary example:
+
+```
+where vec::Iter<let 'a, T>: ExactSizeIterator
+```
+
+Hypothetically, the `let` syntax could be expanded to positions outside of bounds, but
+this RFC proposes no such extension.
+
+## Restrictions on ATCs
+
+In order to be forward compatible with higher order type constructors - which is commonly
+called "full higher kinded types" - this RFC imposes certain restrictions on the kinds of
+constructors which can be used as associated items. Background on this reasoning can be
+found here: http://smallcultfollowing.com/babysteps/blog/2016/11/09/associated-type-constructors-part-4-unifying-atc-and-hkt/
+
+The restriction is simple: Each lifetime argument of the constructor must be applied, in
+order, and must be the left-most arguments of the type constructor. So all of these are
+valid:
+
+```rust
+impl Trait for Type {
+    type Foo<'a> = &'a u32;
+    type Bar<'a> = SomeType<'a, 'static>;
+    type Baz<'a, 'b> = SomeType<'a, 'b>;
+    type Quux<'a> = Self::Foo<'a>;
+}
+```
+
+But these are not valid:
+
+```rust
+impl Trait for Type {
+    type Foo<'a> = String; // ERROR! Argument never used.
+    type Bar<'a> = SomeType<'static, 'a>; // ERROR! Argument must be left-most.
+    type Baz<'a, 'b> = SomeType<'b, 'a>; // ERROR! Arguments are used in wrong order.
+    type Quux<'a> = (&'a i32, &'a i32); // ERROR! Argument is used more than once.
+}
+```
+
+All of these restrictions can be avoided (unpleasantly) with newtypes and phantomdata:
+
+```rust
+struct ValidFoo<'a>(String, PhantomData<&'a ()>);
+struct ValidBar<'a>(SomeType<'static, 'a>);
+struct ValidBaz<'a, 'b>(SomeType<'a, 'b>);
```

Comment:
shouldn't this be:
```rust
struct ValidBaz<'a, 'b>(SomeType<'b, 'a>);
```

(so that then you can feed in `ValidBaz<'a, 'b>` into the original context but get the inner type that we desired?)